subprojects/snapshots/src/main/java/org/gradle/internal/fingerprint/impl/NameOnlyFingerprintingStrategy.java

```diff
@@ -48,39 +48,27 @@ public String normalizePath(CompleteFileSystemLocationSnapshot snapshot) {
 
     @Override
     public Map<String, FileSystemLocationFingerprint> collectFingerprints(Iterable<? extends FileSystemSnapshot> roots) {
-        final ImmutableMap.Builder<String, FileSystemLocationFingerprint> builder = ImmutableMap.builder();
-        final HashSet<String> processedEntries = new HashSet<String>();
+        ImmutableMap.Builder<String, FileSystemLocationFingerprint> builder = ImmutableMap.builder();
+        HashSet<String> processedEntries = new HashSet<>();
         for (FileSystemSnapshot root : roots) {
-            root.accept(new FileSystemSnapshotVisitor() {
-                private boolean root = true;
+            // TODO Use a SnapshotVisitor
+            root.accept(new FileSystemSnapshotHierarchyVisitor() {
+                private boolean seenRoot;
 
                 @Override
-                public boolean preVisitDirectory(CompleteDirectorySnapshot directorySnapshot) {
-                    String absolutePath = directorySnapshot.getAbsolutePath();
+                public SnapshotVisitResult visitEntry(CompleteFileSystemLocationSnapshot snapshot) {
+                    String absolutePath = snapshot.getAbsolutePath();
                     if (processedEntries.add(absolutePath)) {
-                        FileSystemLocationFingerprint fingerprint = isRoot() ? IgnoredPathFileSystemLocationFingerprint.DIRECTORY : new DefaultFileSystemLocationFingerprint(directorySnapshot.getName(), directorySnapshot);
+                        FileSystemLocationFingerprint fingerprint;
+                        if (!seenRoot && snapshot.getType() == FileType.Directory) {
+                            fingerprint = IgnoredPathFileSystemLocationFingerprint.DIRECTORY;
+                            seenRoot = true;
```

Comment:
You need to set `seenRoot = true` outside of `if (processedEntries.add(absolutePath))`. If you don't, `seenRoot` may be wrong when another, earlier snapshot already added part of the current snapshot with a different root. It seems like we are lacking tests in that area. Can you add some?

Reply:
This should be clear now.